src/librustc_mir/transform/promote_consts.rs

```diff
@@ -250,11 +261,11 @@ impl std::ops::Deref for Validator<'a, 'tcx> {
 struct Unpromotable;
 
 impl<'tcx> Validator<'_, 'tcx> {
-    fn validate_candidate(&self, candidate: Candidate) -> Result<(), Unpromotable> {
+    fn validate_candidate(&mut self, candidate: Candidate) -> Result<(), Unpromotable> {
+        self.explicit = candidate.forces_explicit_promotion();
```

Comment:
I didn't want `&mut self`, in order to avoid accidental mutation.
Can you keep the "set outside, assert inside" approach?
With your added method it's much DRY-er.